projects/com.oracle.truffle.llvm.tools/src/com/oracle/truffle/llvm/tools/ToolLookup.java

```diff
@@ -29,42 +29,41 @@
  */
 package com.oracle.truffle.llvm.tools;
 
-import java.io.File;
-
+import com.oracle.truffle.llvm.runtime.LLVMLogger;
 import com.oracle.truffle.llvm.tools.util.ProcessUtil;
-import com.oracle.truffle.llvm.tools.util.ProcessUtil.ProcessResult;
-
-public final class Mx {
-
-    private Mx() {
-    }
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Optional;
+import java.util.function.Predicate;
 
-    public static ProcessResult execute(String string) {
-        String command = "mx " + string;
-        return ProcessUtil.executeNativeCommandZeroReturn(command);
-    }
+final class ToolLookup {
 
-    public static File executeGetLLVMProgramPath(String program) {
-        ProcessResult result = Mx.execute("su-get-llvm " + program);
-        assert result.getReturnValue() == 0;
-        File llvmPath = getProgramPathFromStdout(result);
-        return llvmPath;
-    }
+    private static final int SEARCH_DEPTH = 999;
 
-    public static File executeGetGCCProgramPath(String program) {
-        ProcessResult result = Mx.execute("su-get-gcc " + program);
-        File llvmPath = getProgramPathFromStdout(result);
-        assert result.getReturnValue() == 0;
-        return llvmPath;
+    static Path findTool(String toolNameRegex, Predicate<String> versionPredicate) {
+        String[] pathFolders = System.getenv("PATH").split(":");
+        for (String folder : pathFolders) {
+            try {
+                Optional<Path> findAny = Files.find(Paths.get(folder), SEARCH_DEPTH, (p, f) -> p.getFileName().toFile().getName().matches(toolNameRegex)).filter(p -> versionPredicate.test(getVersion(p.toString()))).findAny();
+                if (findAny.isPresent()) {
+                    return findAny.get();
+                }
+            } catch (IOException | UncheckedIOException ex) {
+                LLVMLogger.info(ex.getMessage());
+            }
+        }
+        return null;
     }
 
-    private static File getProgramPathFromStdout(ProcessResult result) {
-        String output = result.getStdOutput();
-        File llvmPath = new File(output);
-        if (!llvmPath.canExecute()) {
-            throw new AssertionError(output + " is not an executable program!");
+    private static String getVersion(String tool) {
+        try {
+            return ProcessUtil.executeNativeCommand(tool + " --version").getStdOutput();
```

Comment:
I suppose the errors that can occur here and above are due to file permissions. Can we have an explicit check by using `File.canExecute` and throw an AssertionError in the catch block?